### PR TITLE
Favorite tasks

### PR DESCRIPTION
**Pull request by Angatupyry**
## What's new
Favorite tasks are added. 

- The user will be able to create favorite tasks for those he/she considers recurrent.

- The user will be able to edit his/her favorite tasks. 

- The user will be able to delete his/her favorite tasks. 

[chrome-capture-2023-1-13.webm](https://user-images.githubusercontent.com/37310205/218568717-fc308ed7-baf4-46f3-812a-4d18bce12706.webm)

[chrome-capture-2023-1-13 (1).webm](https://user-images.githubusercontent.com/37310205/218568754-d2728195-c134-469e-9479-35ab3f809a9b.webm)

[chrome-capture-2023-1-13 (2).webm](https://user-images.githubusercontent.com/37310205/218568773-045d1e89-c30f-4e86-9c48-2ad2dfbdb0d7.webm)

[chrome-capture-2023-1-13 (3).webm](https://user-images.githubusercontent.com/37310205/218568794-206951c2-ab36-4fc2-9c28-8ff19623348e.webm)

## Self-checks

- [ ] I have prototyped this new feature (if necessary) on Figma 
- [ ] I'm familiar with and follow this [Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [ ] I added unit-tests for new components
- [ ] I tried testing edge cases
- [ ] I tested the behavior of the components that interact with the backend, with an e2e test

### CODE REVIEW

**Review by aaronchongth — CHANGES_REQUESTED** [review]
Thanks for the awesome work, @Angatupyry! This is freaking amazing!

Going back to the issue regarding running on the default sqlite database mentioned in VC, I believe I found the issue, which is that we are not indexing the `user` attribute (which is a `JSONField` and not index-able), I've made some quick suggestions below to change the user field to `CharField` and use `self.user.username` instead of the full json, https://github.com/open-rmf/rmf-web/blob/main/packages/api-server/api_server/models/tortoise_models/user.py#L11. This is under the assumption that all usernames are unique. Some quick testing, the favorite tasks started showing up :partying_face: 

Clicking on the saved favorite task, so far does not populate the create-task-dialog properly for me. If I saved a Delivery task, it correctly populates the task type, but the fields stay empty, similar to the Loop task. This could be due to the changes to `user` field on my end, once those changes have been incorporated, you might see the same issue too. Could you help look into that as well?

File: packages/dashboard/src/components/tasks/tasks-app.tsx [hunk 1]
```diff
@@ -344,6 +396,13 @@ export const TasksApp = React.memo(
               onFail={(e) => {
                 showAlert('error', `Failed to create task: ${e.message}`);
               }}
+              onSuccessFavoriteTask={() => {
+                setOpenCreateTaskForm(false);
+                showAlert('success', 'Successfully');
```
[the remark this answers is not in the document]
Comment (Angatupyry) [reply packages/dashboard/src/components/tasks/tasks-app.tsx]: If the user is creating a favorite task the message will be "Created favorite task successfully", if he/she is deleting the message will be "Deleted favorite task successfully" and if he/she is editing the message will be "Edited favorite task successfully"

File: packages/dashboard/src/components/tasks/tasks-app.tsx [hunk 2]
```diff
@@ -333,9 +382,12 @@ export const TasksApp = React.memo(
               deliveryWaypoints={placeNames}
               dispensers={workcells}
               ingestors={workcells}
+              favoritesTasks={favoritesTasks}
               open={openCreateTaskForm}
               onClose={() => setOpenCreateTaskForm(false)}
               submitTasks={submitTasks}
+              submitFavoriteTask={submitFavoriteTask}
+              deleteFavoriteTask={deleteFavoriteTask}
```
[the remark this answers is not in the document]
Comment (Angatupyry) [reply packages/dashboard/src/components/tasks/tasks-app.tsx:390]: Done! Please take a look!

**Review by aaronchongth — CHANGES_REQUESTED** [review]
Thanks for the quick iteration, it works great! I've left a few nitpicks here and there but the overall feature is working well.

It just occurred to me that there is an important component of Open-RMF that I may not have briefed you about before, which is [rmf_api_msgs](https://github.com/open-rmf/rmf_api_msgs), the models in `api-server/api_server/models/rmf_api` are generated from the json schemas listed in that repository, using this script, https://github.com/open-rmf/rmf-web/blob/main/packages/api-server/generate-models.sh

I would suggest opening a PR on that repository to add the new json schemas for favorite tasks, generate it, and committing those generated models here, as well as updated the commit hash in the generation script. This will ensure compatibility upstream, and that the models we are using, can truly be written in json schema form.

File: packages/react-components/lib/tasks/create-task.tsx [hunk 3]
```diff
@@ -691,6 +789,69 @@ export function CreateTaskForm({
...
+    }
+    setFavoriteTitleError(false);
+
+    return true;
+  };
+
+  const handleSubmitFavoriteTask: React.MouseEventHandler = async (ev) => {
+    ev.preventDefault();
+
+    if (!validateFavoriteTitle()) {
```
[the remark this answers is not in the document]
Comment (Angatupyry) [reply packages/react-components/lib/tasks/create-task.tsx]: I agree with you!

File: packages/react-components/lib/tasks/create-task.tsx [hunk 4]
```diff
@@ -598,13 +679,28 @@ export function CreateTaskForm({
...
   ...otherProps
 }: CreateTaskFormProps): JSX.Element {
   const theme = useTheme();
+
+  const [openFavoriteDialog, setOpenFavoriteDialog] = React.useState(false);
+  const [callToDelete, setCallToDelete] = React.useState(false);
+  const [callToUpdate, setCallToUpdate] = React.useState(false);
+  const [deleting, setDeleting] = React.useState(false);
+
+  const [favoriteTask, setFavoriteTask] = React.useState<TaskFavorite>(defaultFavoriteTask());
```
[the remark this answers is not in the document]
Comment (Angatupyry) [reply packages/react-components/lib/tasks/create-task.tsx:700]: Yeah!
Make sense! I will change it.

File: packages/react-components/lib/tasks/create-task.tsx [hunk 5]
```diff
@@ -691,6 +789,69 @@ export function CreateTaskForm({
...
+    if (!validateFavoriteTitle()) {
+      return;
+    }
+    if (!submitFavoriteTask) {
+      return;
+    }
+
+    setSaving(true);
+    try {
+      setSaving(true);
```
[the remark this answers is not in the document]
Comment (Angatupyry) [reply packages/react-components/lib/tasks/create-task.tsx]: Both are required to momentarily disable the submit button [here](https://github.com/open-rmf/rmf-web/blob/main/packages/react-components/lib/confirmation-dialog.tsx#L74). 
If we merge `setSaving(setSavingFavoriteTask)` and `setDeleting(setDeletingFavoriteTask)` into a single state then that functionality would not be available for the submit button when a user saves a favorite task and then automatically tries to delete some favorite task.

[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/react-components/lib/tasks/create-task.tsx]: Oh sorry I don't mean that, I meant changing
```
setSaving(true);
try {
  setSaving(true);
```
to
```
set Saving(true);
try {
  ...
```

File: packages/api-server/api_server/routes/tasks/tasks.py [hunk 6]
```diff
@@ -217,3 +219,33 @@ async def post_undo_skip_phase(
...
 ):
     return RawJSONResponse(await tasks_service().call(request.json(exclude_none=True)))
+
+
+@router.post(
+    "/favorite_task",
+    response_model=mdl.TaskFavoriteResponseItem,
+    responses={400: {"model": mdl.TaskFavoriteResponseItem1}},
+)
+async def post_favorite_task(
```
[the remark this answers is not in the document]
Comment (Angatupyry) [reply packages/api-server/api_server/routes/tasks/tasks.py:228]: Yeah!
I added some tests.

File: packages/react-components/lib/tasks/create-task.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (Angatupyry) [reply packages/react-components/lib/tasks/create-task.tsx]: Oh, man. I catch it now. Sorry. 
Definitely it is not necessary. Good point. 
Thanks for your review.

File: packages/react-components/lib/tasks/create-task.tsx [hunk 7]
```diff
@@ -14,11 +14,14 @@ import {
   useTheme,
   ListItemIcon,
   IconButton,
+  Typography,
+  ListItemSecondaryAction,
```
Comment (aaronchongth) [inline packages/react-components/lib/tasks/create-task.tsx:18]: nit: could you help sort this in alphabetical order?

**Review by aaronchongth — CHANGES_REQUESTED** [review]
Everything looks good! Just that tiny nitpick regarding import order and we're good to go!

**Comment by aaronchongth** [comment]
Per VC, we will be merging this into `deploy/hammer`, to be backported into `main` after https://github.com/open-rmf/rmf_api_msgs/pull/33 and all other API changes for this project has been merged. This will help us reduce the number of PRs for generation of API messages for the API server, and allow us to fix any incoming broken behaviors once the API is updated.

**Review by aaronchongth — APPROVED** [review]
thanks! LGTM!

**Comment by aaronchongth** [comment]
I noticed that there was a bunch of commits that were not signed earlier in this PR, I don't think it is a big deal, but let's make sure each commit is properly signed with gpg next time. Thanks!

**Pull request merged**